text/0043-better-eviction.md

```diff
@@ -0,0 +1,126 @@
+will be used to initialize the PRNG. That way, an attacker, even if they had the last three blocks
+before the new era's beginning, can choose only among eight different seeds, which are
+all likely to be fair. If they could still e.g. reduce two bids with a million tokens each, they
+could choose among a trillion different proposer sequences. Similarly, if an attacker has k
+validator nodes and could remove themselves by letting any subset of nodes exhibit a fault,
+they could choose among 2<sup>k</sup> leader sequences, even if those k nodes' stakes are small.
+
+We propose a safe way to effectively ignore faulty validators anyway, even if the fault happened
+after the auction, making it easier for the network to remain live without them.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The PRNG seed for an era is still determined by the block in which the auction took place.
+And as a first step to compute the proposer sequence, that seed is still applied to the original
+validator weights.
+
+If after the auction a validator equivocates, it is still set to `Banned` status when the era is
+initialized. In addition, we now also do this if the validator was inactive. I.e. whenever we know
+that a validator is about to be evicted, we mark them as banned.
+
+Now the actual proposer sequence is modified:
+* If the proposer computation with the original set of weights returns a validator that was not
+  banned, the validator retains their slot.
+* If the computation returns a banned validator, the slot is reassigned by running the proposer
+  computation with the modified set of validators, with all banned ones removed.
+
+That way an attacker cannot gain more slots by making their nodes misbehave:
+They only lose some of their own slots.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The first change to the protocol is that in addition to the equivocators, anyone who was reported
+as inactive in a switch block since the auction is marked as banned as well, when a new era is
+initialized. We will apply this only the the past `auction_delay` eras: If someone exhibited a fault
+earlier than that, the auction contract will already have removed them.
+
+The second change is a modification of the proposer slot assignment:
+* We currently add (wrapping, as addition of two unsigned 64-bit integers) the era's seed to the
+  slot's timestamp and seed a
+  [ChaCha PRNG](https://docs.rs/rand_chacha/0.3.0/rand_chacha/struct.ChaCha8Rng.html)
+  with the result, then apply it to select a validator, weighted by their stakes.
+* Now, if the result is a banned validator, we use the same seed, but select a different validator,
+  choosing only among the non-banned ones, weighted by only _their_ stakes.
+
+Thirdly, we remove the banned validators' weights from the total weight for the purpose of fork
```

Comment:
It seems that the proposal here is to, _de facto_, just treat inactive validators as equivocators (without slashing, if it's activated). Is that the case? If so, maybe it would be useful to add an exhaustive description of that case and make it clear that we're just merging the cases.

Reply:
This is only on the consensus side, for the (in mainnet) single era in between, where the inactivity has already been detected but not affected the official validator set as determined by the auction contract yet. This CEP isn't meant to make any recommendation about what to do with equivocators vs. inactive validators later; it just allows us to ignore them before until the auction contract officially removes them.

Because if a validator was completely silent in era 5, we shouldn't expect it to contribute again in era 6; it's safer to remove it from the validator set for fork choice and finality purposes, so the network does not depend on it for liveness or safety at all anymore.